File: Design-Patterns-Using-Go/Mediator/MediatorPattern.go
package main

import "fmt"

// IChatRoom interface
type IChatRoom interface {
	AddParticipant(participant *Participant)
	Broadcast(message, origin string)
	SendMessage(message, to string)
}

// ChatRoom struct implementing IChatRoom
type ChatRoom struct {
	participants map[string]*Participant
}

// AddParticipant adds a participant to the chat room
func (cr *ChatRoom) AddParticipant(participant *Participant) {
	cr.participants[participant.name] = participant
}

// Broadcast sends a message to all participants except the origin
func (cr *ChatRoom) Broadcast(message, origin string) {
	fmt.Printf("ChatRoom broadcast Message: %s\n", message)
	for name, p := range cr.participants {
		if name != origin {
			p.Receive(message)
		}
	}
}

// SendMessage sends a message from one participant to another
func (cr *ChatRoom) SendMessage(message, to string) {
	cr.participants[to].Receive(message)
}

// IParticipant interface
type IParticipant interface {
	Broadcast(message string)
	Send(message, to string)
	Receive(message string)
}

// Participant struct implementing IParticipant
type Participant struct {
	name     string
	chatRoom *ChatRoom
}

// NewParticipant creates a new participant and adds it to the chat room
func NewParticipant(name string, chatRoom *ChatRoom) *Participant {
	participant := &Participant{name: name, chatRoom: chatRoom}
	chatRoom.AddParticipant(participant)
	return participant
}

// Broadcast sends a message to all participants
func (p *Participant) Broadcast(message string) {
	fmt.Printf("%s broadcast Message: %s\n", p.name, message)
	p.chatRoom.Broadcast(message, p.name)
}

// Send sends a message from one participant to another
func (p *Participant) Send(message, to string) {
	fmt.Printf("%s sent Message: %s\n", p.name, message)
	p.chatRoom.SendMessage(message, to)
}

// Receive receives a message
func (p *Participant) Receive(message string) {
	fmt.Printf("%s received Message: %s\n", p.name, message)
}

// Client code
func main() {
	chatRoom := &ChatRoom{participants: make(map[string]*Participant)}
	James := NewParticipant("James", chatRoom)
	Michael := NewParticipant("Michael", chatRoom)
	NewParticipant("Robert", chatRoom)

	Michael.Send("Good Morning.", "James")
	James.Broadcast("Hello, World!")
}

/*
Michael sent Message: Good Morning.
James received Message: Good Morning.
James broadcast Message: Hello, World!
ChatRoom broadcast Message: Hello, World!
Michael received Message: Hello, World!
Robert received Message: Hello, World!
*/

File: Design-Patterns-Using-Swift/Mediator/MediatorPattern.swift
import Foundation

protocol ChatRoom {
    func addParticipant(_ participant: Participant)
    func broadcast(_ message: String, _ origin: String)
    func sendMessage(_ message: String, _ to: String)
}

class ConcreteChatRoom: ChatRoom {
    private var participants = [String: Participant]()

    func addParticipant(_ participant: Participant) {
        participants[participant.getName()] = participant
    }

    func broadcast(_ message: String, _ origin: String) {
        print("ChatRoom broadcast Message : \(message)")
        participants.values.filter { $0.getName() != origin }.forEach { $0.receive(message) }
    }

    func sendMessage(_ message: String, _ to: String) {
        participants[to]?.receive(message)
    }
}

protocol Participant {
    func getName() -> String
    func broadcast(_ message: String)
    func send(_ message: String, _ to: String)
    func receive(_ message: String)
}

class ConcreteParticipant: Participant {
    private let name: String
    private var chatRoom: ChatRoom

    init(name: String, chatRoom: ChatRoom) {
        self.name = name
        self.chatRoom = chatRoom
        chatRoom.addParticipant(self)
    }

    func getName() -> String {
        return name
    }

    func broadcast(_ message: String) {
        print("\(name) broadcast Message : \(message)")
        chatRoom.broadcast(message, name)
    }

    func send(_ message: String, _ to: String) {
        print("\(name) sent Message : \(message)")
        chatRoom.sendMessage(message, to)
    }

    func receive(_ message: String) {
        print("\(name) received Message : \(message)")
    }
}

// Client Code
let chatRoom = ConcreteChatRoom()
let james = ConcreteParticipant(name: "James", chatRoom: chatRoom)
let michael = ConcreteParticipant(name: "Michael", chatRoom: chatRoom)
_ = ConcreteParticipant(name: "Robert", chatRoom: chatRoom)

michael.send("Good Morning.", "James")
james.broadcast("Hello, World!")

/*
Michael sent Message : Good Morning.
James received Message : Good Morning.
James broadcast Message : Hello, World!
ChatRoom broadcast Message : Hello, World!
Robert received Message : Hello, World!
Michael received Message : Hello, World!
*/